language-plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs

```diff
@@ -218,6 +220,73 @@ test_dynamicIdList =
+    toTypeAst _ = case toTypeAst @uni $ Proxy @(TyVarRep text unique) of
+        TyVar () name -> TyForall () name (Type ()) . toTypeAst $ Proxy @a
+        _             -> Prelude.error "Impossible"
+
+dynamicIdRank2Name :: DynamicBuiltinName
+dynamicIdRank2Name = DynamicBuiltinName "idNil"
+
+{- Note [Higher-rank built-in functions]
+We can't unlift a monomorphic function passed to a built-in function, let alone unlift a polymorphic
+one, however we can define a built-in function that accepts an 'Opaque' term of a polymorphic type.
+However, as is always the case with an 'Opaque' term, we can't inspect it or use it in any
+non-opaque way, so a function of type
+
+    all (f :: * -> *). (all (a :: *). f a) -> all (a :: *). f a
+
+can be assigned the following meaning on the Haskell side:
+
+    \x -> x
+
+but we have no way of providing a meaning for a built-in function with the following signature:
+
+    all (f :: * -> *). all (a :: *). (all (a :: *). f a) -> f a
+
+That's because the meaning function would have to instantiate the @all (a :: *). f a@ argument
+somehow to get an @f a@, but that is exactly "using a term in a non-opaque way".
+
+Basically, since we are either generic over @term@ or, like in the example below, use 'CekValue',
+there's is no sensible way of instantiating a passed polymorphic argument (or applying a passed
+argument when it's a function, for another example).
+-}
+
+-- See Note [Higher-rank built-in functions].
+-- >>> :set -XTypeApplications
+-- >>> import Language.PlutusCore.Pretty
+-- >>> putStrLn . render . prettyPlcReadableDef . dynamicBuiltinNameMeaningToType $ dynamicIdRank2Meaning @DefaultUni
+-- (all (f :: * -> *). (all (a :: *). f a) -> (all (a :: *). f a))
+dynamicIdRank2Meaning :: DynamicBuiltinNameMeaning (CekValue uni)
+dynamicIdRank2Meaning = DynamicBuiltinNameMeaning sch id (\_ -> ExBudget 1 1) where
+    sch =
+        TypeSchemeAll @"f" @0 Proxy (KindArrow () (Type ()) $ Type ()) $ \(_ :: Proxy f) ->
+            let ty = Proxy @(Opaque _ (TyForallStarRep "a" 1 (TyAppRep f (Opaque _ (TyVarRep "a" 1)))))
+            in ty `TypeSchemeArrow` TypeSchemeResult ty
+
+dynamicIdRank2Definition :: DynamicBuiltinNameDefinition (CekValue uni)
+dynamicIdRank2Definition =
+    DynamicBuiltinNameDefinition dynamicIdRank2Name dynamicIdRank2Meaning
+
+dynamicIdRank2 :: Term tyname name uni ()
+dynamicIdRank2 = dynamicBuiltinNameAsTerm dynamicIdRank2Name
+
+-- | Test that opaque terms with higher-rank types are allowed.
+test_dynamicIdRank2 :: TestTree
+test_dynamicIdRank2 =
+    testCase "dynamicIdRank2" $ do
+        let env = insertDynamicBuiltinNameDefinition dynamicIdRank2Definition mempty
+            res = mkConstant @Integer @DefaultUni () 0
+            integer = mkTyBuiltin @Integer ()
+            -- sum (idRank2 {list} nil {integer})
+            term
+                = Apply () Plc.sum
+                . TyInst () (Apply () (TyInst () dynamicIdRank2 Plc.listTy) Plc.nil)
+                $ integer
```

Comment:
I got confused by the dots and dollars and thought at first that it was  `(sum (idRank2 {list} nil) {integer}`, which would have been taking the sum of the empty list to obtain a polymorphic value and then instantiating that at the integer type, at which point it magically turned into 0.  That looked extremely improbably, but I see now that I'd misread the term.

However, I did some experimentation and discovered that you could replace `integer` with `unit` and get something that still compiles but which gives you a massive type error at run time.  I guess this is what you were talking about the other day.  I see what you mean: it hadn't occurred to me that this kind of thing could happen!

Reply:
> I got confused by the dots and dollars

Yeah, it takes some time to get used to such definitions.

> `(sum (idRank2 {list} nil) {integer}`

There's a comment saying a different thing (but it could be wrong, of course).

> However, I did some experimentation and discovered that you could replace `integer` with `unit` and get something that still compiles but which gives you a massive type error at run time. I guess this is what you were talking about the other day. I see what you mean: it hadn't occurred to me that this kind of thing could happen!

Wait, no, that would be a true type checker bug. I just trying doing that and got

```
          test/Evaluation/DynamicBuiltins/Definition.hs:288:
          expected: Right (EvaluationSuccess (Constant () (Some (ValueOf integer 0))))
           but got: Left (TypeErrorE (TypeMismatch () <...> <...>))
```

This is not an evaluation error, it's an error that the type checker throws. In particular, `TypeMismatch` is a constructor of `TypeError` and the `TypeErrorE` constructor lifts a `TypeError` into an `Error`.

So it's all fine (apart from the fact that we do not pretty-print errors properly for some reason).

Reply:
Yes, sorry about that. I spotted what was going on after I'd submitted the comment. I should learn to think before typing.

Reply:
No problem, i appreciate that you check things.